include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,407 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_PRECONDITIONER_ILU_HPP_
+#define GKO_CORE_PRECONDITIONER_ILU_HPP_
+
+
+#include <memory>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/std_extensions.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/solver/lower_trs.hpp>
+#include <ginkgo/core/solver/upper_trs.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+#include <ginkgo/core/stop/residual_norm_reduction.hpp>
+
+
+namespace gko {
+namespace preconditioner {
+
+
+/**
+ * The incomplete LU (ILU) preconditioner solves the equation LUx = b for a
+ * given lower triangular matrix L, an upper triangular matrix U and the right
+ * hand side b (can contain multiple right hand sides).
+ *
+ * It allows to set both the solver for L and the solver for U independently,
+ * while providing the defaults solver::LowerTrs and solver::UpperTrs.
+ * For these solvers, a factory can be provided (with `with_l_solver_factory`
+ * and `with_u_solver_factory`) to have more control over their behavior.
+ * In particular, it is possible to use an iterative method for solving the
+ * triangular systems. The default parameters for an iterative triangluar
+ * solver are:
+ * - reduction factor = 1e-4
+ * - max iteration = <number of rows of the matrix given to the solver>
+ * Solvers without such criteria can also be used, in which case none are set.
```

Comment:
I think you can add a mention that by default the solvers are direct triangular solvers. You have mentioned it in the `tparam` description, but I think it would be good to have it in the description as well.

Reply:
Good point, I just added to line 65 (`while providing the defaults solver::LowerTrs and solver::UpperTrs`): `which are direct triangular solvers.`